google/cloud/storage/internal/grpc/metrics_exporter_options.cc

```diff
@@ -12,17 +12,71 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
 #include "google/cloud/storage/internal/grpc/metrics_exporter_options.h"
+#include "google/cloud/opentelemetry/monitoring_exporter.h"
 #include "google/cloud/common_options.h"
+#include "google/cloud/internal/invocation_id_generator.h"
 #include "google/cloud/universe_domain_options.h"
+#include <google/api/monitored_resource.pb.h>
+#include <opentelemetry/sdk/resource/semantic_conventions.h>
+#include <type_traits>
 
 namespace google {
 namespace cloud {
 namespace storage_internal {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+auto ByName(opentelemetry::sdk::resource::ResourceAttributes const& attributes,
+            std::string const& name, std::string default_value) {
+  auto const l = attributes.find(name);
+  if (l == attributes.end()) return default_value;
+  using opentelemetry::nostd::get;
+  return get<std::string>(l->second);
+}
+
+}  // namespace
+
+Options MetricsExporterOptions(
+    Project const& project,
+    opentelemetry::sdk::resource::Resource const& resource,
+    Options const& options) {
+  namespace sc = ::opentelemetry::sdk::resource::SemanticConventions;
+  auto result =
+      Options{}
+          .set<otel_internal::ServiceTimeSeriesOption>(true)
+          .set<otel_internal::MetricPrefixOption>("storage.googleapis.com/");
+
+  // This is fairly expensive, it requires initializing a new PRNG, and fetching
+  // entropy from the OS. Outside tests, this function will be called a handful
+  // of times, so the performance is not that important.
+  auto uuid =
+      google::cloud::internal::InvocationIdGenerator().MakeInvocationId();
+
+  auto const& attributes = resource.GetAttributes();
+  auto monitored_resource = google::api::MonitoredResource{};
+  monitored_resource.set_type("gcs_client_instance");
+  using MapType = std::decay_t<decltype(monitored_resource.labels())>;
+  using Pair = typename MapType::value_type;
+  monitored_resource.mutable_labels()->insert(
+      Pair("project_id", project.project_id()));
+  monitored_resource.mutable_labels()->insert(
+      Pair("location", ByName(attributes, sc::kCloudAvailabilityZone,
+                              ByName(attributes, sc::kCloudRegion, "global"))));
+  monitored_resource.mutable_labels()->insert(Pair(
+      "cloud_platform", ByName(attributes, sc::kCloudPlatform, "unknown")));
+  monitored_resource.mutable_labels()->insert(
+      Pair("host_id", ByName(attributes, "faas.id",
+                             ByName(attributes, sc::kHostId, "unknown"))));
+  monitored_resource.mutable_labels()->insert(
+      Pair("instance_id", std::move(uuid)));
+  monitored_resource.mutable_labels()->insert(Pair("api", "GRPC"));
+
+  result.set<otel_internal::MonitoredResourceOption>(
```

Comment:
Fine for now, but if this option defaulter function becomes more complex, consider factoring out some of the implementation details.

```cc
{
  auto monitored_resource = MakeMonitoredResource(...);
  auto endpoint = CalculateEndpointOption(...);
  return options.set<O1>(...).set<O2>(...).set<O3>(...);
}
```

Reply:
Ack.